pkg/plugins/backendplugin/instrumentation/instrumentation.go

```diff
@@ -58,29 +58,127 @@ const (
 	endpointQueryData      = "queryData"
 )
 
+type errorSource string
+
+const (
+	pluginSource   errorSource = "plugin"
+	externalSource errorSource = "external"
+	databaseSource errorSource = "database"
+	noneSource     errorSource = "none"
+)
+
 var logger = plog.New("plugin.instrumentation")
 
 // instrumentPluginRequest instruments success rate and latency of `fn`
 func instrumentPluginRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, endpoint string, fn func() error) error {
-	status := statusOK
-
 	start := time.Now()
 	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
 
 	err := fn()
-	if err != nil {
-		status = statusError
-		if errors.Is(err, context.Canceled) {
-			status = statusCancelled
+
+	elapsed := time.Since(start)
+	status := getStatus(err)
+
+	logger.Info("Plugin Request Completed")
+
+	updateMetrics(pluginCtx.PluginID, endpoint, string(cfg.Target), elapsed, status, string(noneSource))
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpoint, status, elapsed, timeBeforePluginRequest, err)
+
+	return err
+}
+
+func InstrumentQueryDataRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, requestSize float64, fn func() (*backend.QueryDataResponse, error)) (*backend.QueryDataResponse, error) {
+	start := time.Now()
+	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
+
+	resp, err := fn()
+
+	elapsed := time.Since(start)
+	logger.Info("Plugin Request Completed")
+
+	status := getStatus(err)
+	errorSource := getErrorSource(status, resp)
+
+	updateMetrics(pluginCtx.PluginID, endpointQueryData, string(cfg.Target), elapsed, status, string(errorSource))
+	pluginRequestSizeHistogram.WithLabelValues("grafana-backend", pluginCtx.PluginID, endpointQueryData, string(cfg.Target)).Observe(requestSize)
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpointQueryData, status, elapsed, timeBeforePluginRequest, err)
+
+	return resp, err
+}
+
+func getErrorSourceForResponse(res backend.DataResponse) errorSource {
+	if res.Error != nil {
+		return pluginSource
+	}
+
+	if res.Status >= 500 {
+		return databaseSource
+	}
+
+	if res.Status >= 400 {
+		// Those error codes are related to authentication and authorization.
+		if res.Status == 401 || res.Status == 402 || res.Status == 403 || res.Status == 407 {
+			return externalSource
+		}
+
+		return pluginSource
```

Comment:
Isn't a 400 a client/external source? Client has sent a bad request that plugin couldn't handle.

Reply:
Yes, for the server point of view, 400 means that the client sent invalid data, but we (plugin) are the client here.
I set it to `pluginSource`, but we will probably fine tune that using logs and traces, because those 400 could come from the plugin, who made an error contructing the data, or it might come from the Grafana frontend, which is also owned by us. If the UI that allow this invalid data is from the plugin, then it is pluginSource, if it is from a global Grafana component somewhere, then it could be grafanaSource.

I think defaulting to pluginSource to begin with will allow us to get all our issues, plus other, so it is easier to dispatch in a later step